Firestore/Protos/nanopb_cpp_generator.py

```diff
@@ -246,16 +329,309 @@ def nanopb_write(results):
+      f.content += '''\
+    case %s:
+        return "%s";\n''' % (full_name, short_name)
+
+    f.content += '''
+    }
+    return "<unknown enum value>";
+}\n\n'''
+
+
+def add_printing_for_field(field):
+  """Generates a C++ statement that can print `field` according to its type.
+  """
+  if field.is_optional:
+    return add_printing_for_optional(field)
+  elif field.is_repeated:
+    return add_printing_for_repeated(field)
+  elif field.is_oneof:
+    return add_printing_for_oneof(field)
+  else:
+    return add_printing_for_leaf(field)
+
+
+def add_printing_for_oneof(oneof):
+  """Generates a C++ statement that can print the `oneof` field, if set.
+  """
+  which = oneof.which
+  result = '''\
+    switch (%s) {\n''' % (which)
+
+  for f in oneof.fields:
+    tag_name = '%s_%s_tag' % (oneof.full_classname, f.name)
+    result += '''\
+    case %s:\n''' % tag_name
+
+    # If oneof is set, always print that member, even if it's zero or empty.
+    result += add_printing_for_leaf(f, indent=2, parent_oneof=oneof,
+                                    always_print=True)
+    result += '''\
+        break;\n'''
+
+  result += '''\
+    }\n'''
+
+  return result
+
+
+def add_printing_for_repeated(field):
+  """Generates a C++ statement that can print the repeated `field`, if non-empty.
+  """
+  count = field.name + '_count'
+
+  result = '''\
+    for (pb_size_t i = 0; i != %s; ++i) {\n''' % count
+  # If the repeated field is non-empty, print all its members, even if they are
+  # zero or empty (otherwise, an array of zeroes would be indistinguishable from
+  # an empty array).
+  result += add_printing_for_leaf(field, indent=2, always_print=True)
+  result += '''\
+    }\n'''
+
+  return result
+
+
+def add_printing_for_optional(field):
+  """Generates a C++ statement that can print the optional `field`, if set.
+  """
+  name = field.name
+  result = '''\
+    if (has_%s) {\n''' % name
+  # If an optional field is set, always print the value, even if it's zero or
+  # empty.
+  result += add_printing_for_leaf(field, indent=2, always_print=True)
+  result += '''\
+    }\n'''
+
+  return result
+
+
+def add_printing_for_leaf(field, **kwargs):
+  """Generates a C++ statement that can print the given "leaf" `field`.
+
+  Leaf is to indicate that this function is non-recursive. If `field` is
+  a message, it will delegate printing to its `ToString()` member function.
+  """
+  indent_level = kwargs.get('indent') or 1
+  always_print = 'true' if kwargs.get('always_print') else 'false'
+  parent_oneof = kwargs.get('parent_oneof')
+
+  display_name = field.name
+  cc_name = display_name
+
+  if field.is_repeated:
+    cc_name += '[i]'
+
+  if parent_oneof and not parent_oneof.is_anonymous:
+    cc_name = parent_oneof.name + '.' + cc_name
+
+  if cc_name == 'delete':
+    cc_name = 'delete_'
+  if field.is_primitive:
+    display_name += ':'
+
+  function_name = ''
+  if field.is_enum:
+    function_name = 'PrintEnumField'
```

Comment:
I originally used template metaprogramming for this, but tweaking generated code is probably simpler.